opam build of unison on macOS fails because sys/inotify.h not found

This is on Mac OS X, Version 10.15.7, Build 19H1323.

Here's the build error:
```
$ opam install unison
The following actions will be performed:
  ∗ install unison 2.51.4

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
⬇ retrieved unison.2.51.4  (cached)
[ERROR] The compilation of unison.2.51.4 failed at "dune build -p unison -j 7".

#=== ERROR while compiling unison.2.51.4 ======================================#
# context     2.1.0 | macos/x86_64 | ocaml-base-compiler.4.12.0 | https://opam.ocaml.org#5c5a526e
# path        ~/.opam/forunison/.opam-switch/build/unison.2.51.4
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p unison -j 7
# exit-code   1
# env-file    ~/.opam/log/unison-8482-d96a86.env
# output-file ~/.opam/log/unison-8482-d96a86.out
### output ###
#           cc src/fsmonitor/linux/inotify_stubs.o (exit 1)
# (cd _build/default/src/fsmonitor/linux && /usr/bin/cc -O2 -fno-strict-aliasing -fwrapv -D_FILE_OFFSET_BITS=64 -D_REENTRANT -I/usr/local/opt/expat/include -O2 -fno-strict-aliasing -fwrapv -g -I /Users/tkb/.opam/forunison/lib/ocaml -I ../../lwt -o inotify_stubs.o -c inotify_stubs.c)
# inotify_stubs.c:22:10: fatal error: 'sys/inotify.h' file not found
# #include <sys/inotify.h>
#          ^~~~~~~~~~~~~~~
# 1 error generated.
```

> Do you have XCode or Command Line Tools installed?

Yes, both XCode and the Command Line Tools are installed.  
```
$ xcodebuild -version
Xcode 12.4
Build version 12D4e
```
and
```
$ xcode-select --install
xcode-select: error: command line tools are already installed, use "Software Update" to install updates
```

> Did you try to build it not with opam?

I've now done that, using the ocaml, dune, lablgtk from Homebrew.  
```
$ ocamlopt -version
4.12.0
$ dune --version
2.9.0
```
If I build with `dune build` I get 
```
          cc src/fsmonitor/linux/inotify_stubs.o (exit 1)
(cd _build/default/src/fsmonitor/linux && /usr/bin/cc -O2 -fno-strict-aliasing -fwrapv -D_FILE_OFFSET_BITS=64 -D_REENTRANT -I/usr/local/opt/expat/include -O2 -fno-strict-aliasing -fwrapv -g -I /Users/tkb/.opam/forunison/lib/ocaml -I ../../lwt -o inotify_stubs.o -c inotify_stubs.c)
inotify_stubs.c:22:10: fatal error: 'sys/inotify.h' file not found
#include <sys/inotify.h>
         ^~~~~~~~~~~~~~~
1 error generated.
```

However, if I build it with `make` in the top directory or  `make UISTYLE=gtk2` in the `src` directory it works.  (doing ` make UISTYLE=gtk2` in the top directory doesn't build the gtk2 version, it builds the text UI version.)

> Do you know if previous versions build?

I can build unison-2.51.3 with `make` using the homebrew ocaml if I apply [this](https://github.com/bcpierce00/unison/commit/14b885316e0a4b41cb80fe3daef7950f88be5c8f.patch) patch.  However, when I try to build it it with the homebrew dune I get the following error:
```
       clang src/fsmonitor/linux/inotify_stubs.o (exit 1)
(cd _build/default/src/fsmonitor/linux && /usr/bin/clang -O2 -fno-strict-aliasing -fwrapv -D_FILE_OFFSET_BITS=64 -D_REENTRANT -O2 -fno-strict-aliasing -fwrapv -g -I /usr/local/lib/ocaml -I ../../lwt -o inotify_stubs.o -c inotify_stubs.c)
inotify_stubs.c:30:10: fatal error: 'features.h' file not found
#include <features.h>
         ^~~~~~~~~~~~
1 error generated.
    ocamlopt src/linkgtk2.exe (exit 2)
(cd _build/default && /usr/local/bin/ocamlopt.opt -w @1..3@5..28@30..39@43@46..47@49..57@61..62-40 -strict-sequence -strict-formats -short-paths -keep-locs -w -3-6-9-27-32-52 -g -o src/linkgtk2.exe /usr/local/lib/ocaml/unix.cmxa -I /usr/local/lib/ocaml /usr/local/lib/ocaml/threads/threads.cmxa -I /usr/local/lib/ocaml /usr/local/lib/ocaml/str.cmxa -I /usr/local/lib/ocaml src/lwt/lwt_lib.cmxa /usr/local/lib/ocaml/bigarray.cmxa -I /usr/local/lib/ocaml src/unison_lib.cmxa -I src /Users/tkb/.opam/forunison/lib/lablgtk2/lablgtk.cmxa /Users/tkb/.opam/forunison/lib/lablgtk2/gtkThread.cmx -I /Users/tkb/.opam/forunison/lib/lablgtk2 src/.linkgtk2.eobjs/native/uigtk2.cmx src/.linkgtk2.eobjs/native/linkgtk2.cmx)
Undefined symbols for architecture x86_64:
  "_unsn_hash_univ_param", referenced from:
      _camlUutil__hash_172 in unison_lib.a(uutil.o)
      _camlUutil__1 in unison_lib.a(uutil.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
    ocamlopt src/linktext.exe (exit 2)
(cd _build/default && /usr/local/bin/ocamlopt.opt -w @1..3@5..28@30..39@43@46..4
7@49..57@61..62-40 -strict-sequence -strict-formats -short-paths -keep-locs -g -o src/linktext.exe /usr/local/lib/ocaml/str.cmxa -I /usr/local/lib/ocaml /usr/local/lib/ocaml/unix.cmxa -I /usr/local/lib/ocaml src/lwt/lwt_lib.cmxa /usr/local/lib/ocaml/bigarray.cmxa -I /usr/local/lib/ocaml src/unison_lib.cmxa -I src src/.linktext.eobjs/native/linktext.cmx)
Undefined symbols for architecture x86_64:
  "_unsn_hash_univ_param", referenced from:
      _camlUutil__hash_172 in unison_lib.a(uutil.o)
      _camlUutil__1 in unison_lib.a(uutil.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
File "src/fsmonitor/watchercommon.ml", line 104, characters 8-20:
Error (warning 32 [unused-value-declaration]): unused value really_write.
```
The `unsn_hash_univ_param` undefined symbol errors I take to be part of what the patch mentioned above fixes, but which the dune build doesn't know about, so it is not helpful with figuring out if 2.51.3 can be built with dune.
 
> The build logs from CI might be helpful in figuring this out.

Ok.

I don't know why the dune build is trying to build something that depends on `sys/inotify.h` on macOS, where there is no such file.

--- a/.github/workflows/CICD.yml
+++ b/.github/workflows/CICD.yml
@@ -244,6 +244,7 @@ jobs:
       matrix:
         job:
         - { os: ubuntu-latest  , ocaml-compiler: 4.12.x }
+        - { os: macos-latest  , ocaml-compiler: 4.12.x }
 
     runs-on: ${{ matrix.job.os }}
 

--- a/dune-project
+++ b/dune-project
@@ -1,2 +1,2 @@
-(lang dune 1.3)
+(lang dune 2.3)
 (name unison)

--- a/src/dune
+++ b/src/dune
@@ -7,7 +7,9 @@
         -w -3-6-9-10-26-27-32-34-35-38-39-50-52
         -warn-error -3-6-9-10-26-27-32-34-35-39-50-52
         -no-strict-sequence)
- (c_names bytearray_stubs osxsupport pty hash_compat)
+ (foreign_stubs
+  (language c)
+  (names bytearray_stubs osxsupport pty hash_compat))
  (c_library_flags -lutil)
  (libraries str unix lwt_lib bigarray))
 

--- a/src/fsmonitor/linux/dune
+++ b/src/fsmonitor/linux/dune
@@ -3,14 +3,18 @@
 (library
  (name fswatcher)
  (wrapped false)
+ (enabled_if (= %{system} "linux"))
  (modules :standard \ watcher)
  (flags :standard -w -3-27-39)
- (c_names inotify_stubs)
+ (foreign_stubs
+  (language c)
+  (names inotify_stubs))
  (libraries unix lwt_lib))
 
 (executable
  (name watcher)
  (public_name unison-fsmonitor)
+ (enabled_if (= %{system} "linux"))
  (modules watcher)
  (flags :standard -w -27)
  (libraries fswatcher))

--- a/unison.opam
+++ b/unison.opam
@@ -12,7 +12,7 @@ dev-repo: "git://github.com/bcpierce00/unison.git"
 build: ["dune" "build" "-p" name "-j" jobs]
 depends: [
   "ocaml" {>= "4.03"}
-  "dune" {>= "1.3"}
+  "dune" {>= "2.3"}
   "lablgtk" {>= "2.18.6"}
 ]
 synopsis: "File-synchronization tool for Unix and Windows"

